In [1]:
import pandas as pd
import numpy as np

In [2]:
botComments = pd.read_csv("C:botnet_comments.csv", low_memory=False)

In [3]:
botComments

username  canVote             reatedAt  \
0                                patriot1970    False  2011-04-10T02:19:32   
1                                       qxyz    False  2011-03-02T05:22:53   
2                                   JIM20622    False  2011-02-22T15:06:31   
3                                    theboji    False  2010-12-05T10:59:57   
4                                    theboji    False  2010-12-08T21:44:15   
...                                      ...      ...                  ...   
281352  6th-3b92d6c1a6845bdc4402a0051db060c7    False  2012-04-18T01:54:01   
281353                               Gavinvd    False  2011-02-23T06:03:29   
281354                                 geryd    False  2011-05-23T01:06:48   
281355                         deisymarshall    False  2012-04-02T05:44:40   
281356                           seed7890955    False  2009-01-30T23:25:43   

        dislikes        editableUntil         forum         id  isApproved  \
0              0  2011-04-17T02:19:32  myfoxdetroit  181734313        True   
1              0  2011-03-09T05:22:53  myfoxdetroit  158837937        True   
2              0  2011-03-01T15:06:31  myfoxdetroit  153997782        True   
3              0  2010-12-12T10:59:57  myfoxdetroit  107058805        True   
4              0  2010-12-15T21:44:15  myfoxdetroit  108812146        True   
...          ...                  ...           ...        ...         ...   
281352         0  2012-04-25T01:54:01       allkpop  500382404        True   
281353         0  2011-03-02T06:03:29   kkcooltools  154458299        True   
281354         0  2011-05-30T01:06:48  thedroidguys  209634374        True   
281355         0  2012-04-09T05:44:40     slashfilm  483303469        True   
281356         0  2009-02-06T23:25:43      rankible    5715404        True   

        isAtFlagLimit  isDeleted  ...  isNewUserNeedsApproval  likes  isSpam  \
0               False      False  ...                   False      6   False   
1               False      False  ...                   False      8   False   
2               False      False  ...                   False      4   False   
3               False      False  ...                   False      0   False   
4               False      False  ...                   False      0   False   
...               ...        ...  ...                     ...    ...     ...   
281352          False      False  ...                   False     11   False   
281353          False      False  ...                   False      0   False   
281354          False      False  ...                   False      0   False   
281355          False      False  ...                   False      1   False   
281356          False      False  ...                   False      0   False   

       media                                            message  \
0         []  <p>Someone will definitely be fired over this,...   
1         []  <p>your probably just mad because the only act...   
2         []  <p>I'll probably be hated for this but here go...   
3         []  <p>Pot smokers think weed don't kill, it does,...   
4         []  <p>Looks like scooby snacks is losing weight.</p>   
...      ...                                                ...   
281352    []  <p>congratulations! they've always been legend...   
281353    []  <p>The laser beam is ~0.5cm in diameter at a d...   
281354    []  <p><br>  how does the google meta watch work?</p>   
281355    []  <p> REALLY a wonderful trailer it is i could n...   
281356    []  <p>Thanks for the info. May God have mercy on ...   

        moderationLabels  numReports       parent points     sb  
0                     []         0.0          NaN      6  False  
1            ['flagged']         1.0  158822618.0      8  False  
2                     []         0.0          NaN      4  False  
3                     []         0.0          NaN      0  False  
4            ['flagged']         1.0          NaN      0  False  


In [4]:
botComments = botComments.drop(['numReports', 'sb', 'media', 'moderationLabels', 'editableUntil', 'dislikes', 'canVote', 'isApproved', 'isAtFlagLimit', 'isDeleted', 'isDeletedByAuthor', 'isEdited', 'isFlagged', 'isHighlighted', 'isNewUserNeedsApproval', 'isSpam', 'parent'], axis=1)
botComments

username             reatedAt  \
0                                patriot1970  2011-04-10T02:19:32   
1                                       qxyz  2011-03-02T05:22:53   
2                                   JIM20622  2011-02-22T15:06:31   
3                                    theboji  2010-12-05T10:59:57   
4                                    theboji  2010-12-08T21:44:15   
...                                      ...                  ...   
281352  6th-3b92d6c1a6845bdc4402a0051db060c7  2012-04-18T01:54:01   
281353                               Gavinvd  2011-02-23T06:03:29   
281354                                 geryd  2011-05-23T01:06:48   
281355                         deisymarshall  2012-04-02T05:44:40   
281356                           seed7890955  2009-01-30T23:25:43   

               forum         id  likes  \
0       myfoxdetroit  181734313      6   
1       myfoxdetroit  158837937      8   
2       myfoxdetroit  153997782      4   
3       myfoxdetroit  107058805      0   
4       myfoxdetroit  108812146      0   
...              ...        ...    ...   
281352       allkpop  500382404     11   
281353   kkcooltools  154458299      0   
281354  thedroidguys  209634374      0   
281355     slashfilm  483303469      1   
281356      rankible    5715404      0   

                                                  message  points  
0       <p>Someone will definitely be fired over this,...       6  
1       <p>your probably just mad because the only act...       8  
2       <p>I'll probably be hated for this but here go...       4  
3       <p>Pot smokers think weed don't kill, it does,...       0  
4       <p>Looks like scooby snacks is losing weight.</p>       0  
...                                                   ...     ...  
281352  <p>congratulations! they've always been legend...      11  
281353  <p>The laser beam is ~0.5cm in diameter at a d...       0  
281354  <p><br>  how does the google meta watch work?</p>       0  
281355  <p> REALLY a wonderful trailer it is i could n...       1  
281356  <p>Thanks for the info. May God have mercy on ...       0  

[281357 rows x 7 columns]

***
## Checking which comments appear the most.
***

In [5]:
botComments.message.value_counts()

<p>COMMENT DELETED</p>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  

***
### There was an out of ordinary comment that appeared the third most frequently. It was about Syria so I thought maybe the creator was trying to raise awareness for the country. This could be a starting point to figuring out how these comments were created.
***

In [6]:
syriaSameComments = botComments.loc[botComments.message.str.contains("Please take a few minutes to open your eyes, your mind and your heart ") == True]
syriaComments = botComments.loc[botComments.message.str.contains("Syria") == True]

In [7]:
print(f"""
Difference: {len(syriaComments)} - {len(syriaSameComments)} = {len(syriaComments) - len(syriaSameComments)}
between General Syria comments and Syria comments with the same message
""")


Difference: 6745 - 533 = 6212
between General Syria comments and Syria comments with the same message



### There is not a sizeable difference in number of comments Syria was mentioned vs the most common comment mentioning Syria. There is still a possibility that the bots commenting the same thing could be commenting something different. I will assume that even if the comment is of a different structure, it will still be about Syria, so we will be using syriaComments as our main database.

In [8]:
syriaSameComments

username             reatedAt    forum         id  likes  \
212350  Aleksy_Sokolowski  2012-04-22T05:06:44  ajblogs  505945491      0   
212733    Jaana_Sundqvist  2012-04-19T18:56:47  ajblogs  502624394      0   
218486      Edgar_Lavrinc  2012-04-21T20:58:35  ajblogs  505741454      0   
240077    Atiyeh_Sultanah  2012-04-21T21:49:12  ajblogs  505764890      0   
241659    Felix_Hermansen  2012-04-21T09:24:44  ajblogs  505279634      0   
...                   ...                  ...      ...        ...    ...   
279985     Dhakiyah_Salib  2012-05-11T04:10:43  ajblogs  526035294      0   
279986     Dhakiyah_Salib  2012-05-11T04:10:40  ajblogs  526035275      0   
280276     Ikraam_Sabbagh  2012-05-10T05:41:47  ajblogs  525218132      0   
280374    Mumayyaz_Kattan  2012-05-09T18:56:07  ajblogs  524674984      0   
281335     Qaraja_Mansour  2012-05-11T04:50:27  ajblogs  526052151      0   

                                                  message  points  
212350  <p>Please take a few minutes to open your eyes...       0  
212733  <p>Please take a few minutes to open your eyes...       0  
218486  <p>Please take a few minutes to open your eyes...       0  
240077  <p>Please take a few minutes to open your eyes...       0  
241659  <p>Please take a few minutes to open your eyes...       0  
...                                                   ...     ...  
279985  <p>Please take a few minutes to open your eyes...       0  
279986  <p>Please take a few minutes to open your eyes...       0  
280276  <p>Please take a few minutes to open your eyes...       0  
280374  <p>Please take a few minutes to open your eyes...       0  
281335  <p>Please take a few minutes to open your eyes...       0  

[533 rows x 7 columns]

***
### The similarities in their usernames, in that they're mostly using Pascal(camel+snake) case is possibly indicative of a bot network. These bots could perhaps be working towards raising awareness for Syria. I will chack whether all these comments actually do have the same username formats.
***

In [9]:
len(syriaSameComments.username.str.contains('^[A-Z].+_[A-Z].+', regex=True))

533

### Since the length of this regex filtered variable to only check for Pascal Case is equal to that of the syriaSameComments, we can conclude with 100% certainty that these bots all originated from the same network. We can test the same for general comments about Syria.
***

In [22]:
check = syriaComments.username.str.contains('^[A-Z].+_[A-Z].+', regex=True)
syriaComments = syriaComments[check]
syriaComments

username             reatedAt     forum         id  \
54168               Taxed_Man  2013-01-28T21:16:00  itemlive  781101547   
57512          Amira_Lindgren  2012-03-14T03:34:43   ajblogs  464804949   
57515          Amira_Lindgren  2012-03-14T19:32:47   ajblogs  465360937   
57518          Amira_Lindgren  2012-03-14T19:07:26   ajblogs  465341666   
57524          Amira_Lindgren  2012-03-14T14:57:18   ajblogs  465139946   
...                       ...                  ...       ...        ...   
280534  Jaroslawa_Kwiatkowska  2012-03-13T04:26:51   ajblogs  463988635   
280570          Mikhail_Mahir  2012-03-07T01:58:12   ajblogs  458334866   
280593           Bashair_Harb  2012-03-12T23:41:34   ajblogs  463820217   
281322       Medoro_Langelier  2012-03-13T04:48:21   ajblogs  463998358   
281335         Qaraja_Mansour  2012-05-11T04:50:27   ajblogs  526052151   

        likes                                            message  points  
54168      15  <p>The unarmed people of Syria are being murde...      15  
57512       0  <p>Tweeties, be careful what you ask for! Why ...       0  
57515       0  <p>Do you know why the murderer, rapists and a...       0  
57518       0  <p>Every Syrian will be forever so grateful fo...       0  
57524       0  <p>RobotNick,<br>The number of messages on the...       0  
...       ...                                                ...     ...  
280534      0  <p>Listen to the people you are cheering askin...       0  
280570      0  <p>James Hunter, Joris Diepshidd, Tony Two-Fac...       0  
280593      0  <p>How about we Syrians, all Syrians, not just...       0  
281322      0  <p>Save Syria! Say NO to Ibn Saud!</p><blockqu...       0  
281335      0  <p>Please take a few minutes to open your eyes...       0  

[5018 rows x 7 columns]

***
### Here we can see that there are 5018 potential bot users comments with the same username format with different comment structures about Syria.
***
On a side-note, I see that the user truefaith1 which made the most comments overall, also had comments on Syria.
For instance,

In [11]:
syriaComments.loc[1426].message

"<p>In other news...ISIS might be making a comeback.  Not surprising since ISIS was odumdum's JV team.  Leaves me wondering now who he thought was his varsity team...was it the Clowns In America?  And they used the JV team to provide distraction and cover?  And what are they distracting from and covering up THIS time?  And why are we bombing Syria again?  Lord, I pray that the truth will be revealed.  And in such a way that it cannot be denied or covered or ignored.  Please show the serpent for the little worm he is now.  Thank You Father in Heaven that the truth WILL prevail and that You have already won the war!</p>"

In [12]:
botComments.username.value_counts()

truefaith1            4836
KayakMomma            4298
912anita              3710
vipcassieblackjack    3463
looserbigbang         2713
                      ... 
disqus_4RjbWRjJto        1
AngelicaN                1
finesse183               1
disqus_SAMn2KBvGY        1
seed7890955              1
Name: username, Length: 29578, dtype: int64

### One possible explanation for this bot commenting on Syria is because it is a controversial topic. There could be an intent behind this content but I cannot know for sure currently.

### What I can do though is figure out how many of the comments from syriaComments are unique.

In [23]:
syriaComments.username.value_counts()

Amira_Lindgren             152
Martim_Dreadnaught          68
Tahan_Sakhr                 23
Kalb_Aroob                  14
Sabine_Rothschild           14
                          ... 
Crassus_Goldworthy           1
Wakil_Almasi                 1
Bogumil_Stojanovic           1
Andreja_Marija_Rendulic      1
Qaraja_Mansour               1
Name: username, Length: 1198, dtype: int64

### There are a total of 1198 unique bot users from these comments, all with the same Pascal name format. I think this is a bot network. I will try to confirm by matching these usernames with the password_ids from our potential bots dataset.

In [27]:
potentialBots = pd.read_csv("C:potential_social_botnets.csv")
potentialBots

username        email_id   email_provider_id  \
0         Eleanor2345   email_9794418   email_provider_14   
1         Carling2345   email_9794410   email_provider_14   
2        Cecania23456   email_9794246   email_provider_14   
3       Courtland2345   email_9794196   email_provider_14   
4            Arlo1234   email_9926096  email_provider_292   
...               ...             ...                 ...   
363940     Joukosmith  email_14485231    email_provider_0   
363941   Vincencsmith   email_7850375    email_provider_0   
363942     Kaleosmith  email_11405041    email_provider_0   
363943  Filibertsmith   email_5964889    email_provider_0   
363944    Osheensmith  email_15803421    email_provider_1   

        is_major_email_provider       email_reuse_id       password_id  \
0                          True     email_reuse_6100  password_1003131   
1                          True     email_reuse_6100  password_1003131   
2                          True     email_reuse_6100  password_1003131   
3                          True     email_reuse_6100  password_1003131   
4                          True     email_reuse_6101  password_1003131   
...                         ...                  ...               ...   
363940                     True  email_reuse_4156781   password_999325   
363941                     True  email_reuse_4156782   password_999325   
363942                     True  email_reuse_4156783   password_999325   
363943                     True  email_reuse_4156784   password_999325   
363944                     True  email_reuse_4156785   password_999325   

        is_common_password                  creation_date  
0                    False  2009-11-14 00:41:42.176683-05  
1                    False  2009-11-11 04:27:49.997487-05  
2                    False  2009-11-11 07:15:30.413369-05  
3                    False  2009-11-13 02:36:45.308864-05  
4                    False  2009-11-06 06:39:56.472807-05  
...                    ...                            ...  
363940               False  2011-11-23 06:09:10.295889-05  
363941               False  2011-11-18 01:19:11.701083-05  
363942               False  2012-02-01 23:27:54.085126-05  
363943               False  2011-09-16 02:51:13.792899-04  
363944               False  2012-01-05 23:58:57.145921-05  

[363945 rows x 8 columns]

### Extracting Pascal case usernames from potential bots.

In [40]:
potentialBots = potentialBots[potentialBots.username.str.contains('^[A-Z].+_[A-Z].+', regex=True)]

In [41]:
potentialBots.head(10)

username        email_id  email_provider_id  \
932         Clatan_Surgeon    email_901408  email_provider_14   
4372       Gabriela_Brownv    email_658372   email_provider_0   
18436           Patty_Melt  email_12844195  email_provider_14   
19254          Zach_Tanner   email_7660554   email_provider_1   
19296                Mi_St  email_11493732  email_provider_14   
27165       Isabella_Texas  email_13446676  email_provider_14   
27166    Hannah_Washington   email_3006756  email_provider_14   
27167         Lily_Georgia  email_11956084  email_provider_14   
27168          Ava_Georgia  email_11956085  email_provider_14   
27169  Payton_Rhode_Island   email_3006757  email_provider_14   

       is_major_email_provider      email_reuse_id       password_id  \
932                       True   email_reuse_13849   password_100736   
4372                      True  email_reuse_116458  password_1052496   
18436                     True  email_reuse_334738  password_1150915   
19254                     True  email_reuse_349706  password_1157297   
19296                     True  email_reuse_349748  password_1157297   
27165                     True  email_reuse_445212  password_1204215   
27166                     True  email_reuse_445213  password_1204215   
27167                     True  email_reuse_445214  password_1204215   
27168                     True  email_reuse_445214  password_1204215   
27169                     True  email_reuse_445213  password_1204215   

       is_common_password                  creation_date  
932                 False  2011-08-17 11:33:10.924926-04  
4372                False  2011-05-17 04:37:46.597983-04  
18436               False  2011-05-18 13:00:44.093858-04  
19254               False   2010-12-15 21:42:29.08946-05  
19296               False  2010-03-22 08:02:06.648648-04  
27165               False  2012-02-01 04:18:29.631742-05  
27166               False  2012-01-31 20:12:41.962406-05  
27167               False  2012-01-28 10:28:41.576063-05  
27168               False  2012-01-28 12:57:29.203751-05  
27169               False  2012-02-01 10:01:43.855473-05

### Matching syriaComments unique usernames to potentialBots unique usernames to get password_id for each.

In [55]:
allBots = potentialBots[potentialBots.username.isin(set(syriaComments.username.unique()))]
allBots.shape

(1197, 8)

In [58]:
abs(len(allBots.username) - len(syriaComments.username.value_counts()))

1

### There is only a difference of one between all potential bots with Pascal username format and bots that commented about Syria, which means we are only missing one user. I am even more certain now that these bots are working together. I want to compare password_ids now.

In [62]:
allBots.password_id.value_counts()

password_423651    1196
password_72079        1
Name: password_id, dtype: int64

### There is an outlier bot user but we can disregard that.
# Conclusion: This network of bots sharing Pascal name formats and commenting about Syria are working together, as derived from the identical password_ids (password_423651). Their goal could be to raise awareness on the situation in Syria. It could also be that they are commenting on something controversial to gain points or I'm not sure, but we saw this intent in the bot named truefaith1 that made most comments overall also mentioning Syria.